flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/MemoryArchivist.scala

```diff
@@ -25,48 +25,82 @@ import org.apache.flink.runtime.jobgraph.JobID
 import org.apache.flink.runtime.messages.ArchiveMessages._
 import org.apache.flink.runtime.messages.JobManagerMessages._
 
+import scala.collection.mutable.LinkedHashMap
+import scala.ref.SoftReference
+
 class MemoryArchivist(private val max_entries: Int) extends Actor with ActorLogMessages with
 ActorLogging {
   /**
    * Map of execution graphs belonging to recently started jobs with the time stamp of the last
-   * received job event.
+   * received job event. The insert order is preserved through a LinkedHashMap.
    */
-  val graphs = collection.mutable.HashMap[JobID, ExecutionGraph]()
-  val lru = collection.mutable.Queue[JobID]()
+  val graphs = LinkedHashMap[JobID, SoftReference[ExecutionGraph]]()
 
   override def receiveWithLogMessages: Receive = {
+    /* Receive Execution Graph to archive */
     case ArchiveExecutionGraph(jobID, graph) => {
-      graphs.update(jobID, graph)
+      // wrap graph inside a soft reference
+      graphs.update(jobID, new SoftReference(graph))
+
+      // clear all execution edges of the graph
+      val iter = graph.getAllExecutionVertices().iterator()
+      while (iter.hasNext) {
+        iter.next().clearExecutionEdges()
+      }
+
       cleanup(jobID)
     }
 
     case RequestArchivedJobs => {
-      sender ! ArchivedJobs(graphs.values)
+      sender ! ArchivedJobs(getAllGraphs())
     }
 
     case RequestJob(jobID) => {
-      graphs.get(jobID) match {
-        case Some(graph) => sender ! JobFound(jobID, graph)
-        case None => sender ! JobNotFound(jobID)
+      getGraph(jobID) match {
```

Comment:
Ah I see, because of the soft ref. Why not letting getGraph return an Option type instead of null. Null values are evil. [Tony Hoares multi-billion dollar mistake] (http://en.wikipedia.org/wiki/Tony_Hoare#Quotations)